lib/bolt/node/winrm.rb

```diff
@@ -72,17 +72,38 @@ def shell_init
   $startInfo = New-Object System.Diagnostics.ProcessStartInfo($Path, $Arguments)
   $startInfo.UseShellExecute = $false
   $startInfo.WorkingDirectory = Split-Path -Parent (Get-Command $Path).Path
+  $startInfo.CreateNoWindow = $true
   if ($StdinInput) { $startInfo.RedirectStandardInput = $true }
   $startInfo.RedirectStandardOutput = $true
   $startInfo.RedirectStandardError = $true
 
+  $stdoutHandler = { if (-not ([String]::IsNullOrEmpty($EventArgs.Data))) { $Host.UI.WriteLine($EventArgs.Data) } }
+  $stderrHandler = { if (-not ([String]::IsNullOrEmpty($EventArgs.Data))) { $Host.UI.WriteErrorLine($EventArgs.Data) } }
+  $invocationId = [Guid]::NewGuid().ToString()
+
   try
   {
-    $process = [System.Diagnostics.Process]::Start($startInfo)
+    $process = New-Object System.Diagnostics.Process
+    $process.StartInfo = $startInfo
+    $process.EnableRaisingEvents = $true
+
+    # https://msdn.microsoft.com/en-us/library/system.diagnostics.process.standarderror(v=vs.110).aspx#Anchor_2
+    $stdoutEvent = Register-ObjectEvent -InputObject $process -EventName 'OutputDataReceived' -Action $stdoutHandler
+    $stderrEvent = Register-ObjectEvent -InputObject $process -EventName 'ErrorDataReceived' -Action $stderrHandler
+    $exitedEvent = Register-ObjectEvent -InputObject $process -EventName 'Exited' -SourceIdentifier $invocationId
+
+    $process.Start() | Out-Null
+
+    $process.BeginOutputReadLine()
+    $process.BeginErrorReadLine()
   }
   catch
   {
-    Write-Error $_
+    $Host.UI.WriteErrorLine($_)
+    @($stdoutEvent, $stderrEvent, $exitedEvent) |
+      ? { $_ -ne $Null } |
+      % { Unregister-Event -SourceIdentifier $_.Name }
+    if ($process -ne $Null) { $process.Dispose() }
```

Comment:
What about moving the `Unregister-Event` and `Dispose` logic into a finally clause so that it doesn't need to be repeated below?

```powershell
try {
    ...
    return $process.ExitCode
} catch {
    ...
   return 1
} finally {
    @($stdoutEvent, $stderrEvent, $exitedEvent) | ? { $_ -ne $Null } | % { Unregister-Event -SourceIdentifier $_.Name }
    $process.Dispose()
}
```

Reply:
Had thought about that, but was concerned with muddying the logic around the process kill.  I agree the duplication is not ideal.

Let me see if I can cleanly refactor / restructure that

Reply:
Logic has been reworked into a single try-catch-finally, which required restructuring the termination... but intent should be the same (aside from additionally terminating runaway processes under other failure scenarios)